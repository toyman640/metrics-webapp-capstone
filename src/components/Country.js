import React, { useEffect } from 'react';
import { useSelector } from 'react-redux';
import { useParams, useNavigate } from 'react-router-dom';

function Country() {
  const countries = useSelector((state) => state.countries.countrydata);

  const { id } = useParams();
  const index = countries.findIndex((country) => country.id === +id);
  const navigate = useNavigate();

  useEffect(() => {
    const handleBeforeUnload = (event) => {
      event.preventDefault();
      const confirmationMessage = '';
      return confirmationMessage;
    };
    const handleBeforeUnloadWrapper = (event) => {
      handleBeforeUnload(event);
      navigate('/');
    };
    window.addEventListener('beforeunload', handleBeforeUnloadWrapper);
    return () => {
      window.removeEventListener('beforeunload', handleBeforeUnloadWrapper);
    };
  }, [navigate]);

  if (!countries || countries.length === 0 || index === -1) {
    return <div>Loading...</div>;
  }

  return (
    <div className="Detail-page">
      <div
        className="Country-detial"
        style={{
          backgroundImage: `url(${countries[index].coat})`,
          backgroundSize: 'contain',
          backgroundRepeat: 'no-repeat',
          position: 'relative',
        }}
      />
      <div className="Data-section">
        <h3 className="Details-name">
          Country Name:
          {' '}
          <span className="">{countries[index].name}</span>
        </h3>
        <h3 className="Details-name">
          Population:
          {' '}
          <span>{countries[index].population}</span>
        </h3>
        <h3 className="Details-name">
          Capital:
          {' '}
          <span>{countries[index].capital}</span>
        </h3>
        <h3 className="Details-name">
          Subregion:
          {' '}
          <span>{countries[index].subregion}</span>
        </h3>
        <h3 className="Details-name">
          Timezone:
          {' '}
          <span>{countries[index].timezone}</span>
        </h3>
      </div>
    </div>
  );
}

export default Country;
