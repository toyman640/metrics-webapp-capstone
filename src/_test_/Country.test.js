import React from 'react';
import { render } from '@testing-library/react';
import { useSelector } from 'react-redux';
import { BrowserRouter as Router } from 'react-router-dom';

import Country from '../components/Country';

jest.mock('react-redux', () => ({
  useSelector: jest.fn(),
}));

describe('Country', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders loading message when countries data is not available', () => {
    useSelector.mockReturnValue([]);

    const { container } = render(
      <Router>
        <Country />
      </Router>,
    );

    expect(container.textContent).toBe('Loading...');
  });

  it('renders country details when countries data is available', () => {
    const countriesData = [
      {
        id: 1,
        coat: 'example-coat-url',
        name: 'Country 1',
        population: 1000000,
        capital: 'Capital 1',
        subregion: 'Subregion 1',
        timezone: 'Timezone 1',
      },
    ];

    useSelector.mockReturnValue(countriesData);

    const { container } = render(
      <Router>
        <Country />
      </Router>,
    );

    expect(container).toMatchSnapshot();
  });
});
