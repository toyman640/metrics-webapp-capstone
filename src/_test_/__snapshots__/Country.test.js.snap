// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Country renders country details when countries data is available 1`] = `
<div>
  <div>
    Loading...
  </div>
</div>
`;
